lib/revolution.rb

```
@@ -1,9 +1,127 @@
+#!/usr/bin/env ruby
 # frozen_string_literal: true
 
 require 'revolution/version'
+require 'revolution/build'
+require 'revolution/deploy'
+require 'revolution/order'
+require 'revolution/recipes'
+require 'thor'
 
 module Revolution
-  def self.run
-    puts 'Running!'
+  # TODO: separate CLI class from main Revolution logic
+  class CLI < Thor
+    desc 'all', 'Start Revolution'
+    option :recipe_root, required: true
+    def all
+      puts 'Doing all the things!'
+      diff
+      build
+      sign
+      deploy
+    end
+
+    desc 'diff', 'Check for packages that have changed'
+    option :recipe_root, required: true
+    def diff
+      puts 'Checking for package updates...'
+      # TODO
+    end
+
+    desc 'order', 'Get build order based on dependencies'
+    option :recipe_root, required: true
+    def order
+      puts 'Getting build order...'
+      return StandardError unless valid_dir?(options.recipe_root)
+      recipes        = Recipes.load_recipes(options.recipe_root)
+      sorted_recipes = Order.resolve_build_order(recipes)
+      output_build_order(sorted_recipes)
+      sorted_recipes
+    end
+
+    desc 'build', 'Build packages in order according to dependencies'
+    option :recipe_root, required: true
+    # TODO: add --all option, package_name parameter
+    def build
+      puts 'Building packages in order...'
+      sorted_recipes = order # gets options.recipe_root implicitly
+      built_packages = []
+      sorted_recipes.each do |recipe|
+        # TODO: Add option for verbose output
+        puts "Building package: #{recipe.package_name}"
+        package_path  = options.recipe_root + recipe.package_name
+        is_clean      = Build.clean(package_path)
```

Instead of doing long chains of "is_clean" or "is_successful" like this... maybe if the clean fails it should raise an error?